compiler/rustc_next_trait_solver/src/solve/effect_goals.rs

```diff
@@ -225,11 +229,104 @@ where
     }
 
     fn consider_builtin_fn_trait_candidates(
-        _ecx: &mut EvalCtxt<'_, D>,
-        _goal: Goal<I, Self>,
+        ecx: &mut EvalCtxt<'_, D>,
+        goal: Goal<I, Self>,
         _kind: rustc_type_ir::ClosureKind,
     ) -> Result<Candidate<I>, NoSolution> {
-        todo!("Fn* are not yet const")
+        let cx = ecx.cx();
+
+        let self_ty = goal.predicate.self_ty();
+        let (tupled_inputs_and_output, def_id, args) = match self_ty.kind() {
```

Comment:
We're not reusing the helper from `structural_traits` b/c we specifically want to extract a `DefId` here too. 

I guess alternatively we could pass back an `Option<DefId>` from that helper and bail if it's `None` (i.e. a `FnPtr`), then do the const check afterwards.